packages/adapters-library/src/defiProvider.ts

```diff
@@ -800,6 +803,36 @@ export class DefiProvider {
     filterProtocolTokens?: string[]
     blockNumbers?: Partial<Record<Chain, number>>
   }): Promise<TotalValueLockResponse[]> {
+    this.validateTvlInputs({
+      filterProtocolIds,
+      filterProductIds,
+      filterChainIds,
+      filterProtocolTokens,
+    })
+
+    const chainId = filterChainIds![0]!
+    const protocolAddress = filterProtocolTokens![0]!
+    const protocolId = filterProtocolIds![0]!
+    const productId = filterProductIds![0]!
+
+    const adapter = await this.adaptersController.fetchAdapter(
+      chainId,
+      protocolId,
+      productId,
+    )
+
+    if (!adapter) {
+      throw new TvlValidationError(
+        `No adapter found for protocol address: ${protocolAddress} on chain: ${chainId}`,
+      )
+    }
+
+    if (adapter.protocolId !== protocolId || adapter.productId !== productId) {
+      throw new TvlValidationError(
+        `Protocol address ${protocolAddress} adapter does not match the provided protocolId: ${protocolId} and productId ${productId}`,
+      )
+    }
+
```

Comment:
Previous implementation was spamming price adapter.

Now instead first we find the correct adapter and run tvl only for that adapter.